src/main/java/io/nats/client/impl/NatsMessageConsumer.java

```diff
@@ -52,43 +52,50 @@ class NatsMessageConsumer extends NatsMessageConsumerBase implements PullManager
     @Override
     public void heartbeatError() {
         try {
-            // just close the current sub and make another one.
-            // this could go on endlessly - unless the user had called stop
             if (stopped.get()) {
-                finishAndClose();
+                finishAndShutdownSub();
             }
             else {
-                lenientClose();
+                shutdownSub();
                 doSub();
             }
         }
         catch (JetStreamApiException | IOException e) {
-            pmm.resetTracking();
-            pmm.initOrResetHeartbeatTimer();
+            setupHbAlarmToTrigger();
         }
     }
 
     void doSub() throws JetStreamApiException, IOException {
         MessageHandler mh = userMessageHandler == null ? null : msg -> {
             userMessageHandler.onMessage(msg);
             if (stopped.get() && pmm.noMorePending()) {
-                finishAndClose();
+                finished.set(true);
```

Comment:
This matches before I, ahem, "fixed" (broke) it
https://github.com/nats-io/nats.java/blob/0e4f39c569e7db99cf116651f7d63e41c87d73d6/src/main/java/io/nats/client/impl/NatsMessageConsumer.java